feat: add `block.timestamp` asserter for AA

ischasny opened the pull request:
This PR adds the ability to use `block.timestamp` in custom AA contracts. AAs will still not have direct access to `block.timestamp`, but can utilize it via a proxy that enforces certain constraints. The PR introduces a `TimestampAsserter` contract that is deployed on every chain to the user space, similar to `Multicall3`. This contract has a single function, `assertTimestampInRange(start, end)`, which can be used by AAs at their discretion.

The `TimestampAsserter` contract ensures that `block.timestamp` falls within the specified `(start, end)` range. Additionally, the sequencer verifies that the `block.timestamp` is sufficiently far from the range’s end. This is to prevent DoS attacks where transactions pass validation but get stuck in the mempool during execution. This constraint is configurable and can be adjusted without requiring protocol update.

The PR also introduces two new fields to the `transactions` table: `timestamp_asserter_range_start` and `timestamp_asserter_range_end`. These fields are extracted during transaction execution in the sandbox by the `ValidationTracer`. If multiple assertions are made in a single transaction, the system captures the maximum of the starts and the minimum of the ends, resulting in the narrowest possible time range.

Transactions with time range constraints will undergo additional verification before being included in a block. If the current time falls outside the transaction’s specified time range, the transaction will be rejected with an appropriate message.

Sister PR in `era-contracts`: https://github.com/matter-labs/era-contracts/pull/843

slowli commented on core/node/state_keeper/src/io/mempool.rs:
```diff
@@ -298,6 +298,31 @@ impl StateKeeperIO for MempoolIO {
+                // Reject transactions that violate block.timestamp constraints. Such transactions should be
+                // rejected at the API level, but we need to protect ourselves in case if a transaction
+                // goes outside of the allowed range while being in the mempool
+                let now = SystemTime::now()
```
Same question as for the VM tracer: why is time acquired here instead of using one from the latest block?

slowli commented on core/node/state_keeper/src/io/mempool.rs:
```diff
@@ -298,6 +298,31 @@ impl StateKeeperIO for MempoolIO {
                     continue;
                 }
+
+                // Reject transactions that violate block.timestamp constraints. Such transactions should be
+                // rejected at the API level, but we need to protect ourselves in case if a transaction
+                // goes outside of the allowed range while being in the mempool
```
Dumb question: The PR description says that changing these params doesn't require a protocol upgrade. How will params work in a decentralized network? How will these values be coordinated among validators? (I'd imagine the timestamp asserter contract would actually contain these values in this case.) Or is this too far too ahead to plan?

ischasny replied:
We could in theory move these checks down to the `TimestampAsserter` contract and have those variables set there. @StanislavBreadless  would love your thoughts here.

StanislavBreadless replied:
I think it is better not to do things on the protocol side that can be moved somewhere else. (In the worst case, consensus can be done offchain). 

The reason why communicating it offchain may make sense is because this restriction is on the API side only. I.e. if someone proposes a miniblock, the validators that actually check the block should not be affected by this restriction, i.e. if the validation of the custom account passes, the block is valid. Otherwise it is not. The restrictions here are just to protect the block builder and not the other validators

slowli commented on core/node/api_server/src/tx_sender/tests/send_tx.rs:
```diff
@@ -298,3 +308,75 @@ async fn sending_transaction_out_of_gas() {
 }
+
+#[tokio::test]
+async fn submitting_tx_with_validation_traces() {
```
Bikeshedding: IMO, this test is mostly useless; it essentially tests that the mock tx executor works as expected. I'd advise to test the *real* executor (as most tests in this module do) with the changed custom account contract; so essentially, move the new integration tests here. Likewise, you can test gas estimation as well.

slowli commented on core/node/api_server/src/execution_sandbox/validate.rs:
```diff
@@ -39,13 +39,16 @@ impl SandboxExecutor {
             &tx,
             self.options.eth_call.validation_computational_gas_limit(),
             whitelisted_tokens_for_aa,
+            self.timestamp_asserter_address,
+            self.timestamp_asserter_min_range_sec,
+            self.timestamp_asserter_min_time_till_end_sec,
```
IIUC, during node synchronization it can still accept transaction requests (i.e., validate transactions). If this is the case, is the fact that validation can be called with `timestamp_asserter_address == Some(_)` before the asserter contract is deployed a concern?

ischasny commented on core/lib/dal/src/transactions_dal.rs:
```diff
@@ -347,7 +357,9 @@ impl TransactionsDal<'_, '_> {
                 execution_info,
                 received_at,
                 created_at,
-                updated_at
+                updated_at,
+                timestamp_asserter_range_start,
+                timestamp_asserter_range_end
```
Good point - fixed

ischasny commented on core/lib/dal/src/models/storage_transaction.rs:
```diff
@@ -321,6 +324,15 @@ impl From<StorageTransaction> for Transaction {
     }
 }
 
+impl From<StorageTransaction> for TransactionTimeRangeConstraint {
+    fn from(tx: StorageTransaction) -> Self {
+        Self {
+            timestamp_asserter_range_start: tx.timestamp_asserter_range_start,
+            timestamp_asserter_range_end: tx.timestamp_asserter_range_end,
+        }
+    }
+}
```
Fixed

ischasny commented on core/lib/types/src/lib.rs:
```diff
@@ -416,3 +417,19 @@ impl Transaction {
+    Copy
+)]
+pub struct TransactionTimeRangeConstraint {
+    pub timestamp_asserter_range_start: Option<NaiveDateTime>,
+    pub timestamp_asserter_range_end: Option<NaiveDateTime>,
```
Changed to `i64`

ischasny commented on core/lib/multivm/src/tracers/validator/vm_latest/mod.rs:
```diff
@@ -81,6 +86,56 @@ impl<H: HistoryMode> ValidationTracer<H> {
+                    // If this is a call to the timestamp asserter, extract the function arguments and store them in ValidationTraces.
+                    // These arguments are used by the mempool for transaction filtering. The call data length should be 68 bytes:
+                    // a 4-byte function selector followed by two U256 values.
+                    // TODO: if any more functions will be added to the TimestampAsserter contract in the future, verify the function selector too
```
Good point - added

ischasny commented on core/lib/multivm/src/tracers/validator/vm_latest/mod.rs:
```diff
@@ -81,6 +86,56 @@ impl<H: HistoryMode> ValidationTracer<H> {
+                    // These arguments are used by the mempool for transaction filtering. The call data length should be 68 bytes:
+                    // a 4-byte function selector followed by two U256 values.
+                    // TODO: if any more functions will be added to the TimestampAsserter contract in the future, verify the function selector too
+                    if Some(called_address) == self.timestamp_asserter_address
+                        && far_call_abi.memory_quasi_fat_pointer.length == 68
```
The reason I implemented it that way - was because that's the way it's done right above for [KECCAK](https://github.com/matter-labs/zksync-era/blob/main/core/lib/multivm/src/tracers/validator/vm_latest/mod.rs#L52). 

I tried doing in the way you suggested above - calldata gets red always as zeroes. Does any modificaiton need to be done to the snippet you provided?

ischasny commented on core/lib/multivm/src/tracers/validator/vm_latest/mod.rs:
```diff
@@ -81,6 +86,56 @@ impl<H: HistoryMode> ValidationTracer<H> {
+                            return Err(ViolatedValidationRule::TimestampAssertionShortRange);
+                        }
+
+                        let now = SystemTime::now()
```
Good point - fixed

ischasny commented on core/lib/multivm/src/tracers/validator/vm_latest/mod.rs:
```diff
@@ -81,6 +86,56 @@ impl<H: HistoryMode> ValidationTracer<H> {
+
+                        {
+                            let mut traces_mut = self.traces.lock().unwrap();
+                            traces_mut.timestamp_asserter_range_start =
```
Reworked it to be `Option<std::ops::RangeInclusive<_>>` - like.

ischasny commented on core/lib/vm_interface/src/types/tracer.rs:
```diff
@@ -57,6 +57,13 @@ pub struct ValidationParams {
     pub computational_gas_limit: u32,
+    /// Address of the timestamp asserter. This contract is allowed to touch block.timestamp regardless
+    /// of the calling context.
+    pub timestamp_asserter_address: Option<Address>,
+    /// Minimum difference in seconds between the range start and range end
+    pub timestamp_asserter_min_range_sec: u32,
+    /// Minimum time between current block.timestamp and the end of the asserted range
+    pub timestamp_asserter_min_time_till_end_sec: u32,
```
Good point - changed

ischasny commented on core/node/api_server/src/execution_sandbox/execute.rs:
```diff
@@ -100,13 +100,19 @@ pub(crate) struct SandboxExecutor {
     engine: SandboxExecutorEngine,
     pub(super) options: SandboxExecutorOptions,
     storage_caches: Option<PostgresStorageCaches>,
+    pub(super) timestamp_asserter_address: Option<Address>,
+    pub(super) timestamp_asserter_min_range_sec: u32,
+    pub(super) timestamp_asserter_min_time_till_end_sec: u32,
```
Fixed

ischasny commented on core/node/api_server/src/execution_sandbox/execute.rs:
```diff
@@ -100,13 +100,19 @@ pub(crate) struct SandboxExecutor {
         options: SandboxExecutorOptions,
         caches: PostgresStorageCaches,
         missed_storage_invocation_limit: usize,
+        timestamp_asserter_address: Option<Address>,
+        timestamp_asserter_min_range_sec: u32,
+        timestamp_asserter_min_time_till_end_sec: u32,
```
Grouped them up into a single struct - let me know if it looks better now.

slowli commented on core/lib/multivm/src/tracers/validator/vm_latest/mod.rs:
```diff
@@ -81,6 +83,50 @@ impl<H: HistoryMode> ValidationTracer<H> {
+                            }
+
+                            if end.as_u32()
+                                < self.l1_batch_env.timestamp as u32 + params.min_time_till_end_sec
```
I think it could be worth clarifying in a comment why using `l1_batch_env.timestamp` is OK here. AFAIU, it is because this tracer is always used in a oneshot execution mode (executing a single transaction).

ischasny replied:
indeed that's the case - added clarification

slowli commented on core/lib/vm_interface/src/types/errors/halt.rs:
```diff
@@ -42,6 +42,7 @@ pub enum Halt {
     VMPanic,
     TracerCustom(String),
     FailedToPublishCompressedBytecodes,
+    FailedBlockTimestampAssertion,
```
Future work (?): AFAICT, right now this error is only raised outside the VM, by `MempoolIO`. Is it possible to convert actual `require` failure in the timestamp asserter to this variant? Or would it require too much work?

ischasny replied:
Thanks. Created a ticket for future investigation.

slowli commented on core/lib/vm_interface/src/types/tracer.rs:
```diff
@@ -104,6 +127,29 @@ pub enum ValidationError {
+    /// Merges two ranges together by taking the maximum of the starts and the minimum of the ends
+    /// resulting into the narrowest possible time window
+    pub fn apply_range(&mut self, new_range: (i64, i64)) {
+        if let Some(mut range) = self.timestamp_asserter_range {
```
This copies the range, which is probably not what you want to do here. To modify the range in place, use `if let Some(range) = &mut self.timestamp_asserter_range`. I'd suggest to add a couple of unit tests to check that this method works as intended.

ischasny replied:
Great point - changed and added tests

slowli commented on core/node/api_server/src/tx_sender/mod.rs:
```diff
@@ -368,7 +390,7 @@ impl TxSender {
             .0
             .tx_sink
-            .submit_tx(&tx, execution_output.metrics)
+            .submit_tx(&tx, execution_output.metrics, validation_result.unwrap())
```
Please replace the `let Err(_) = validation_result` to make it clear that this is safe:

```rust
let validation_traces = validation_result?;
// ...
.submit_tx(&tx, execution_output.metrics, validation_traces)
```

slowli commented on core/node/api_server/src/tx_sender/mod.rs:
```diff
@@ -234,6 +247,9 @@ pub struct TxSenderConfig {
     pub validation_computational_gas_limit: u32,
     pub chain_id: L2ChainId,
     pub whitelisted_tokens_for_aa: Vec<Address>,
+    pub timestamp_asserter_address: Option<Address>,
+    pub timestamp_asserter_min_range_sec: u32,
+    pub timestamp_asserter_min_time_till_end_sec: u32,
```
Why not use `Option<TimestampAsserterParams>` here? Do `min_range_sec` / `min_time_till_end_sec` params still make sense even if the asserter address isn't set?

ischasny replied:
Makes sense - changed

slowli commented on core/node/state_keeper/Cargo.toml:
```diff
@@ -31,6 +31,8 @@ zksync_node_test_utils.workspace = true
 zksync_vm_executor.workspace = true
 zksync_system_constants.workspace = true
 zksync_base_token_adjuster.workspace = true
+chrono = { workspace = true, features = ["serde"] }
+
```
This dependency seems unused as well.

ischasny replied:
Removed

slowli commented on core/node/state_keeper/src/io/mempool.rs:
```diff
@@ -298,6 +299,23 @@ impl StateKeeperIO for MempoolIO {
+                // rejected at the API level, but we need to protect ourselves in case if a transaction
+                // goes outside of the allowed range while being in the mempool
+                let matches_range = constraint.timestamp_asserter_range.map_or(true, |x| {
+                    x.0 < l2_block_timestamp as i64 && x.1 > l2_block_timestamp as i64
```
Bikeshedding: IMO, a range with both exclusive ends is somewhat esoteric, even not taking Rust into account (there's no easy way to construct such ranges, unlike `x..y` / `x..=y`). If I'd read the `assertTimestampInRange` name, I'd assume that both range ends are inclusive. I understand that this is very subjective though.

ischasny replied:
Refined the logic with `range.contains` instead

slowli commented on core/node/state_keeper/src/io/tests/mod.rs:
```diff
@@ -644,3 +666,121 @@ async fn insert_unsealed_batch_on_init(commitment_mode: L1BatchCommitmentMode) {
+            timestamp_asserter_range: Some((system_time - 1000, system_time + 1000)),
+        },
+    );
+    #[allow(deprecated)]
```
Dumb question: What's deprecated in this statement? It seems to work fine if the attribute is removed.

ischasny replied:
Thanks, good spot - it's a leftover from previous iterations where I used `NaiveDateTime` type

slowli commented on core/node/state_keeper/src/io/tests/mod.rs:
```diff
@@ -644,3 +666,121 @@ async fn insert_unsealed_batch_on_init(commitment_mode: L1BatchCommitmentMode) {
+    let tx = mempool
+        .wait_for_next_tx(Duration::from_secs(2), system_time as u64)
+        .await
+        .expect("No expected transaction in the mempool")
+        .unwrap();
```
Nit: judging by `expect` messages, `expect` should be switched around with `unwrap()`; i.e., `unwrap` should be applied to a `Result`, and `expect` to the underlying `Option`. Ditto below.

ischasny replied:
Fixed for all, thanks

slowli commented on core/tests/ts-integration/contracts/custom-account/custom-account.sol:
```diff
@@ -9,6 +9,10 @@ import './SystemContractsCaller.sol';
 import './interfaces/IAccount.sol';
 
+interface ITimestampAsserter {
+	function assertTimestampInRange(uint256 start, uint256 end) external view;
```
Bikeshedding: I think (fortunately), tab indentation isn't usually used in Solidity. Would be somewhat to change in this case though.

ischasny replied:
Do you mean changing TAB to 4 spaces?

slowli commented on core/tests/ts-integration/tests/custom-account.test.ts:
```diff
@@ -95,6 +113,94 @@ describe('Tests for the custom account behavior', () => {
+            from: await customAccount.getAddress()
+        });
+
+        // requires custom assertion because the transaction would fail in "estimateFee" that is not handled correctly by zksync ethers
```
Dumb question: Would it be possible to use another RPC method then? (E.g., submit a transaction w/o estimating gas for it.)

ischasny replied:
You'd still have to fill all required fields though no? Even if you call 'sendRawTransaction` without estimating fee first.

slowli replied:
Yes, but I thought that in this case it would be possible to simplify error assertions below similarly to other tests.

ischasny replied:
Very good point - changed the test

slowli commented on core/tests/ts-integration/tests/custom-account.test.ts:
```diff
@@ -95,6 +113,94 @@ describe('Tests for the custom account behavior', () => {
 
+    test('Should fail transaction validation due to timestamp assertion in the validation tracer - short time window', async () => {
+        const now = Math.floor(Date.now() / 1000);
+        const rangeStart = now - 60;
+        const rangeEnd = now + 60;
```
How do we know that this range is too short? Maybe, it would make sense to read node config and then create a shorter range? Ditto for the next test.

slowli commented on core/tests/ts-integration/tests/custom-account.test.ts:
```diff
@@ -95,6 +113,94 @@ describe('Tests for the custom account behavior', () => {
+                fail('The transaction was expected to fail due to timestamp assertion');
+            })
+            .catch((error) => {
+                // If it's an error object with the structure you're mentioning, check its properties
```
Nit: Who's "you"? 🙃

slowli commented on core/tests/ts-integration/tests/custom-account.test.ts:
```diff
@@ -95,6 +113,94 @@ describe('Tests for the custom account behavior', () => {
+                        'failed to validate the transaction. reason: Timestamp is out of range'
+                    );
+                } else {
+                    fail('The transaction was expected to fail with a different message');
```
Please use `expect(null).fail(..)` as in other integration tests.

ischasny replied:
Fixed

slowli commented on core/tests/ts-integration/tests/custom-account.test.ts:
```diff
@@ -40,7 +45,18 @@ describe('Tests for the custom account behavior', () => {
+        customAccount = await deployContract(
+            alice,
+            contracts.customAccount,
+            // 2555971200 is a number of seconds up to 30/12/2050
```
Nit: This comment should be moved to the `DEFAULT_TIMESTAMP_ASSERTER_RANGE_END` declaration.

ischasny commented on core/lib/config/src/configs/chain.rs:
```diff
@@ -244,3 +244,11 @@ impl MempoolConfig {
 }
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Default)]
+pub struct TimestampAsserterConfig {
```
I think it's a general pattern if you look into how configs are structured across the board. Agree that it'd make sense to enforce reasonable defaults everywhere if such exist, but if I do this for one specific config it'd rather become a weird outlier.

ischasny commented on core/lib/protobuf_config/src/timestamp_asserter.rs:
```diff
@@ -0,0 +1,19 @@
+    type Type = TimestampAsserterConfig;
+    fn read(&self) -> anyhow::Result<Self::Type> {
+        Ok(Self::Type {
+            min_range_sec: *required(&self.min_range_sec).unwrap_or(&0),
+            min_time_till_end_sec: *required(&self.min_time_till_end_sec).unwrap_or(&0),
```
Fixed - errors are propagated back up now.

ischasny commented on core/lib/types/src/lib.rs:
```diff
@@ -416,3 +416,19 @@ impl Transaction {
+)]
+pub struct TransactionTimeRangeConstraint {
+    /// Represents a range from-to. Each field is a number of seconds since the epoch.
+    pub timestamp_asserter_range: Option<(i64, i64)>,
```
Changed to Range<u64> for both;

ischasny commented on core/lib/dal/migrations/20241008073940_add_block_timestamp_asserter.down.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE transactions DROP COLUMN timestamp_asserter_range_start;
+ALTER TABLE transactions DROP COLUMN timestamp_asserter_range_end;
```
Fixed

ischasny commented on core/lib/dal/src/transactions_dal.rs:
```diff
@@ -1819,8 +1837,14 @@ impl TransactionsDal<'_, '_> {
+        let transactions_with_constraints = transactions
+            .into_iter()
+            .map(|tx| {
+                let constraint: TransactionTimeRangeConstraint = (&tx).into();
```
Applied

ischasny commented on core/lib/mempool/Cargo.toml:
```diff
@@ -13,3 +13,5 @@ categories.workspace = true
 [dependencies]
 zksync_types.workspace = true
 tracing.workspace = true
+chrono = { workspace = true, features = ["serde"] }
+
```
Good spot - removed

ischasny commented on core/lib/multivm/src/tracers/validator/vm_latest/mod.rs:
```diff
@@ -81,6 +83,50 @@ impl<H: HistoryMode> ValidationTracer<H> {
+
+                            let params = self.timestamp_asserter_params.as_ref().unwrap();
+
+                            if end.as_u32() - start.as_u32() < params.min_range_sec {
```
* `end` is guaranteed to be greater than `start`, this is enforced in the contract, i.e.
```
        require(start < block.timestamp && end > block.timestamp, "Timestamp is out of range");
```
* added an extra bit of logic to avoid overflows - values are capped at `u64::MAX` / `u32::MAX`

joonazan commented on core/lib/multivm/src/tracers/validator/vm_latest/mod.rs:
```diff
@@ -81,6 +83,50 @@ impl<H: HistoryMode> ValidationTracer<H> {
+                    // These arguments are used by the mempool for transaction filtering. The call data length should be 68 bytes:
+                    // a 4-byte function selector followed by two U256 values.
+                    let address = self.timestamp_asserter_params.as_ref().map(|x| x.address);
+                    if Some(called_address) == address
```
I'd write something like`if self.timestamp_asserter_params.map(|x| x.address == called_address).unwrap_or_default()`

joonazan replied:
Actually, to avoid the `.unwrap()` later on, you should do `if let Some(address) = ...` here.

ischasny commented on core/lib/types/src/lib.rs:
```diff
@@ -416,3 +416,19 @@ impl Transaction {
+
+#[derive(
+    Clone,
+    Serialize,
+    Deserialize,
```
Indeed - removed.

joonazan commented:
I believe that the "time range is wide enough" check does nothing and should be removed. Suppose you want the time range [1, 2]. You can get it by requiring two ranges: [-500, 2] and [1, 500].

The check that the range's end is far enough in the future is sufficient to prevent ultra-short ranges, I think.

ischasny commented:
> if Some(called_address) == address

Great point thanks. Indeed we don't need to check how wide the range is. Just making sure that:
* `block.timestamp` is within the specified range (done by the contract);
* and that the end is enough in the future 
should be sufficient.

Removed the check and updated the PR description.

slowli commented on core/node/state_keeper/src/io/tests/mod.rs:
```diff
@@ -644,3 +666,118 @@ async fn insert_unsealed_batch_on_init(commitment_mode: L1BatchCommitmentMode) {
+#[tokio::test]
+async fn test_mempool_with_timestamp_assertion(commitment_mode: L1BatchCommitmentMode) {
+    let connection_pool = ConnectionPool::<Core>::constrained_test_pool(2).await;
+    let tester = Tester::new(commitment_mode);
```
Nit: I don't think the commitment mode matters here, so it would be sufficient to set it to any single value.

slowli commented on core/node/state_keeper/src/mempool_actor.rs:
```diff
@@ -122,16 +123,22 @@ impl MempoolFetcher {
+            let transactions = transactions_with_constraints
+                .iter()
+                .map(|(t, _c)| t)
+                .collect_vec();
```
Bikeshedding: It's easy to avoid dependency on `Itertools` by using `collect` with a type annotation or a turbofish.

ischasny replied:
changed

slowli commented on core/node/state_keeper/src/io/mod.rs:
```diff
@@ -137,8 +137,11 @@ pub trait StateKeeperIO: 'static + Send + Sync + fmt::Debug + IoSealCriteria {
+    async fn wait_for_next_tx(
+        &mut self,
+        max_wait: Duration,
+        l2_block_timestamp: u64,
```
If it's guaranteed that `wait_for_new_batch_params` / `wait_for_new_l2_block_params` output is used for L2 blocks, the L2 block timestamp can be encapsulated inside `MempoolIO`, and this method can be left unchanged.

ischasny replied:
I chose to change the interface because `UpdatesManager` gets initialised (outside)[https://github.com/matter-labs/zksync-era/blob/main/core/node/state_keeper/src/keeper.rs#L151] of the wiring framework. Getting it encapsulated into the `MempoolIO` would require larger refactoring.

slowli commented on core/tests/ts-integration/tests/custom-account.test.ts:
```diff
@@ -95,6 +112,68 @@ describe('Tests for the custom account behavior', () => {
+                testMaster.environment().l2ChainId
+            )
+        ).toBeRejected(
+            'failed to validate the transaction. reason: Violated validation rules: block.timestamp is too close to the range end'
```
Nit: Would match via a regex work here (like `/block\.timestamp/`)? It's usually a bad idea to tie tests to human-readable error messages.

ischasny replied:
Agreed - this is rather a pattern that is used across all integration tests. Ideally i'd match against a defined set of error codes.

slowli commented on zkstack_cli/crates/zkstack/completion/_zkstack.zsh:
Dumb question: What are changes in this file caused by? It doesn't look like this PR changes `zkstack` command-line args.

ischasny replied:
Not sure - the PR started failing lint check in the CI - maybe job config has changed

ischasny replied:
FYI - tagged you in relevant slack thread - probably an unformatted file has made it's way into the main somehow.

ischasny commented on core/lib/dal/src/transactions_dal.rs:
```diff
@@ -346,6 +357,8 @@ impl TransactionsDal<'_, '_> {
                 paymaster_input,
                 execution_info,
                 received_at,
+                timestamp_asserter_range_start,
+                timestamp_asserter_range_end,
```
There was a discussion about that in the internal notion - @RomanBrodetski agreed to keep it as separate fields.

ischasny commented on core/node/api_server/src/tx_sender/mod.rs:
```diff
@@ -234,6 +244,14 @@ pub struct TxSenderConfig {
+}
+
+#[derive(Debug, Clone)]
+pub struct TimestampAsserterParams {
```
Because it won't work in the external node as it's not aware of that crate.

slowli replied:
It can be re-exported from the API server (the simplest option, probably). Conceptually, there would be no overhead adding a direct dependency for the EN either because it is included in its dependency graph anyway.

ischasny commented on core/bin/external_node/src/config/mod.rs:
```diff
@@ -685,6 +698,16 @@ impl OptionalENConfig {
+                .timestamp_asserter_config
+                .as_ref()
+                .map(|x| x.min_range_sec)
+                .unwrap_or(0),
```
Changed to use more reasonable default

ischasny commented on core/lib/multivm/src/tracers/validator/mod.rs:
```diff
@@ -60,8 +67,14 @@ impl<H> ValidationTracer<H> {
+    ) -> (
+        Self,
+        Arc<OnceCell<ViolatedValidationRule>>,
+        Arc<Mutex<ValidationTraces>>,
```
Added

ischasny commented on core/lib/vm_interface/src/types/tracer.rs:
```diff
@@ -104,6 +127,30 @@ pub enum ValidationError {
+    /// a lack of overlap would have triggered an assertion failure in the `TimestampAsserter` contract,
+    /// as `block.timestamp` cannot satisfy two non-overlapping ranges.
+    pub fn apply_range(&mut self, new_range: Range<u64>) {
+        if let Some(ref mut range) = self.timestamp_asserter_range.as_mut() {
```
Changed

ischasny commented on core/lib/vm_interface/src/types/tracer.rs:
```diff
@@ -116,3 +163,45 @@ impl fmt::Display for ValidationError {
+        let mut validation_traces = ValidationTraces {
+            timestamp_asserter_range: None,
+        };
+        let new_range = Range { start: 10, end: 20 };
```
changed

ischasny commented on core/lib/dal/src/models/storage_transaction.rs:
```diff
@@ -321,6 +324,30 @@ impl From<StorageTransaction> for Transaction {
+            Self {
+                timestamp_asserter_range: Some(
+                    tx.timestamp_asserter_range_start
+                        .unwrap()
```
Simplified it in a bit different way

ischasny commented on core/lib/dal/src/transactions_dal.rs:
```diff
@@ -2212,6 +2237,29 @@ impl TransactionsDal<'_, '_> {
         .await
     }
+
+    pub async fn get_storage_tx_by_hash(
```
I agree - I wouldn't be asserting a error message if we've had some standardised failure reason.

slowli replied:
Can (IMO should) be fixed in a follow-up PR.

ischasny commented on core/lib/vm_interface/src/types/tracer.rs:
```diff
@@ -57,6 +57,19 @@ pub struct ValidationParams {
+    /// Minimum difference in seconds between the range start and range end
+    pub min_range_sec: u32,
+    /// Minimum time between current block.timestamp and the end of the asserted range
+    pub min_time_till_end_sec: u32,
```
Good point - changed

ischasny commented on core/bin/zksync_server/src/node_builder.rs:
```diff
@@ -322,6 +323,13 @@ impl MainNodeBuilder {
+                    .map(|address| TimestampAsserterParams {
+                        address,
+                        min_range_sec: timestamp_asserter_config.min_range_sec,
+                        min_time_till_end_sec: timestamp_asserter_config.min_time_till_end_sec,
```
I'd be ok with leaving it as is, but also happy to change it in a follow up PR if you feel strong about it.

slowli commented on core/lib/protobuf_config/src/timestamp_asserter.rs:
```diff
@@ -0,0 +1,19 @@
+    type Type = TimestampAsserterConfig;
+    fn read(&self) -> anyhow::Result<Self::Type> {
+        Ok(Self::Type {
+            min_time_till_end_sec: *required(&self.min_time_till_end_sec)
```
Don't forget to add this config param to stage / other auto-deployed envs; IIUC, they'll get broken otherwise.

slowli commented on core/bin/external_node/src/config/mod.rs:
```diff
@@ -146,6 +147,10 @@ impl RemoteENConfig {
             .get_main_contract()
             .rpc_context("get_main_contract")
             .await?;
+        let timestamp_asserter_address = client
```
Is there a risk that the main node won't have this method implemented (e.g., on the mainnet)? If so, it makes sense to handle `MethodNotFound` errors as it's done below.

ischasny replied:
Agreed and fixed

slowli commented on core/bin/external_node/src/config/mod.rs:
```diff
@@ -1447,6 +1467,14 @@ impl From<&ExternalNodeConfig> for TxSenderConfig {
+                TimestampAsserterParams {
+                    address,
+                    min_time_till_end: Duration::from_secs(
+                        config.optional.timestamp_asserter_min_time_till_end_sec as u64,
```
Nit: Since `u32` -> `u64` conversion is lossless, you may use `timestamp_asserter_min_time_till_end_sec.into()` to signal it. Or change the field type to `u64`.

slowli commented on core/lib/dal/src/transactions_dal.rs:
```diff
@@ -2212,6 +2238,29 @@ impl TransactionsDal<'_, '_> {
+            hash.as_bytes()
+        )
+        .map(Into::into)
+        .instrument("get_tx_error_by_hash")
```
Nit: Mismatch between the method name / semantics and the name here. As we've discussed, I'd prefer if the method actually just loaded the error btw 🙃

slowli commented on core/lib/multivm/src/tracers/validator/vm_latest/mod.rs:
```diff
@@ -81,6 +83,52 @@ impl<H: HistoryMode> ValidationTracer<H> {
+                                {
+                                    let mut traces_mut = self.traces.lock().unwrap();
+                                    traces_mut.apply_range(start..end);
+                                }
```
Nit: Since the outer scope ends here anyway, it's not necessary to create an additional scope for the mutex guard. Also, since it's used for a single operation, it's possible to apply it directly w/o declaring a local var.

slowli commented on core/lib/vm_interface/src/types/tracer.rs:
```diff
@@ -104,6 +120,30 @@ pub enum ValidationError {
+    /// producing the narrowest possible time window. Note that overlapping ranges are essential;
+    /// a lack of overlap would have triggered an assertion failure in the `TimestampAsserter` contract,
+    /// as `block.timestamp` cannot satisfy two non-overlapping ranges.
+    pub fn apply_range(&mut self, new_range: Range<u64>) {
```
Bikeshedding: If we plan to add new fields to `ValidationTraces` in the future, this method name / docs look too ambiguous. It would make sense to call it like `(apply|narrow)_timestamp_asserter_range`.

slowli commented on core/tests/ts-integration/tests/custom-account.test.ts:
```diff
@@ -192,19 +192,41 @@ describe('Tests for the custom account behavior', () => {
+            const calldataMatch = err.message.match(/data\s=\s(0x[0-9a-fA-F]+)/);
+
+            if (!functionSelectorMatch || !calldataMatch) {
+                expect(null).fail('Function selector or calldata not found in the error message');
```
Nit: You can use [`toBeTruthy`](https://jestjs.io/docs/expect#tobetruthy) matcher here to be more idiomatic. Also, `!` below would fulfill the same goal (albeit without an error message).

slowli commented on core/tests/ts-integration/tests/custom-account.test.ts:
```diff
@@ -192,19 +192,41 @@ describe('Tests for the custom account behavior', () => {
+            expect(err.message).toContain(
+                'failed to validate the transaction. reason: Validation revert: Account validation error'
+            );
+            const functionSelectorMatch = err.message.match(/function_selector\s=\s(0x[0-9a-fA-F]{8})/);
+            const calldataMatch = err.message.match(/data\s=\s(0x[0-9a-fA-F]+)/);
```
Off-topic, probably: Parsing function selector / calldata from the human-readable message looks weird. Is this what Web3 clients expect? I'd imagine an optimal implementation would define separate fields in the JSON-RPC response.

ischasny replied:
Agreed - I'd prefer custom errors to be returned as a part of JSON RPC response.

slowli commented on core/tests/ts-integration/tests/custom-account.test.ts:
```diff
@@ -192,19 +192,41 @@ describe('Tests for the custom account behavior', () => {
+            const start = BigInt(`0x${startHex}`);
+            const end = BigInt(`0x${endHex}`);
+
+            expect(functionSelector).toBe('0x3d5740d9');
```
Nit: I'd move this next to the `functionSelector` declaration; if this assertion doesn't hold, parsing `start` / `end` may fail leading to hard-to-attribute errors.

slowli reviewed (commented):
Other than a potential error getting remote EN config, looks fine.

slowli commented on core/bin/external_node/src/config/mod.rs:
```diff
@@ -227,10 +227,36 @@ impl RemoteENConfig {
+    F: Future<Output = Result<T, ClientError>>,
+    T: Clone,
+{
+    match rpc_call.await {
```
Nit: You can match the call result directly instead of `await`ing for it.

slowli commented on core/bin/external_node/src/config/mod.rs:
```diff
@@ -227,10 +227,36 @@ impl RemoteENConfig {
+            ]
+            .contains(&(err.code())) =>
+        {
+            fallback.ok_or_else(|| anyhow::anyhow!("RPC method not found or internal error"))
```
I don't understand how `fallback == None` works; as it's implemented currently, it's not a fallback at all, and a node executing this code will still break down here unless the main node is updated. Isn't `None` a valid value for the `timestamp_asserter_address`? If so, it looks like `fallback` should have type `T` and it should be returned here unconditionally.

ischasny replied:
Good point - fixed

slowli commented on core/bin/zksync_server/src/node_builder.rs:
```diff
@@ -303,6 +305,20 @@ impl MainNodeBuilder {
+                Some(TimestampAsserterParams {
+                    address,
+                    min_time_till_end: Duration::from_secs(
+                        timestamp_asserter_config.min_time_till_end_sec as u64,
```
Nit: This lossless conversion can be replaced with `_.into()` as well.

yorik commented on .github/workflows/new-build-core-template.yml:
```diff
@@ -208,6 +211,14 @@ jobs:
+      - name: Create necessary artefacts
+        shell: bash
+        run: |
+          mkdir -p contracts/l1-contracts/artifacts/
```
nit: you can create all the directories with single command:
`mkdir -p contracts/l1-contracts/artifacts contracts/l1-contracts/out`

yorik reviewed (dismissed):
Approve for workflows.
